plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAccessControl.java

```diff
@@ -0,0 +1,1145 @@
+    {
+        opaHighLevelClient.queryAndEnforce(
+                OpaQueryContext.fromSystemSecurityContext(context),
+                "CreateSchema",
+                AccessDeniedException::denyCreateSchema,
+                schema,
+                properties);
+    }
+
+    @Override
+    public void checkCanDropSchema(SystemSecurityContext context, CatalogSchemaName schema)
+    {
+        opaHighLevelClient.queryAndEnforce(
+                OpaQueryContext.fromSystemSecurityContext(context),
+                "DropSchema",
+                AccessDeniedException::denyDropSchema,
+                schema);
+    }
+
+    @Override
+    public void checkCanRenameSchema(SystemSecurityContext context, CatalogSchemaName schema, String newSchemaName)
+    {
+        OpaQueryInputResource resource = new OpaQueryInputResource
+                .Builder()
+                .schema(TrinoSchema.fromTrinoCatalogSchema(schema))
+                .build();
+        OpaQueryInputResource targetResource = new OpaQueryInputResource
+                .Builder()
+                .schema(new TrinoSchema.Builder()
+                        .catalogName(schema.getCatalogName())
+                        .schemaName(newSchemaName)
+                        .build())
+                .build();
+
+        OpaQueryContext queryContext = OpaQueryContext.fromSystemSecurityContext(context);
+
+        if (!opaHighLevelClient.queryOpaWithSourceAndTargetResource(queryContext, "RenameSchema", resource, targetResource)) {
+            denyRenameSchema(schema.toString(), newSchemaName);
+        }
+    }
+
+    @Override
+    public void checkCanSetSchemaAuthorization(SystemSecurityContext context, CatalogSchemaName schema, TrinoPrincipal principal)
+    {
+        OpaQueryInputResource resource = new OpaQueryInputResource
+                .Builder()
+                .schema(TrinoSchema.fromTrinoCatalogSchema(schema))
+                .build();
+        OpaQueryInputGrant grantee = new OpaQueryInputGrant
+                .Builder()
+                .principal(TrinoGrantPrincipal.fromTrinoPrincipal(principal))
+                .build();
+        OpaQueryInputAction action = new OpaQueryInputAction
+                .Builder()
+                .operation("SetSchemaAuthorization")
+                .resource(resource)
+                .grantee(grantee)
+                .build();
+        OpaQueryInput input = new OpaQueryInput(OpaQueryContext.fromSystemSecurityContext(context), action);
+
+        if (!opaHighLevelClient.queryOpa(input)) {
+            denySetSchemaAuthorization(schema.toString(), principal);
+        }
+    }
+
+    @Override
+    public void checkCanShowSchemas(SystemSecurityContext context, String catalogName)
+    {
+        opaHighLevelClient.queryAndEnforce(
+                OpaQueryContext.fromSystemSecurityContext(context),
+                "ShowSchemas",
+                (DenyCallable) AccessDeniedException::denyShowSchemas,
```

Comment:
This `queryAndEnforce` argument is a callable that will be invoked if authorization is denied by OPA.
`queryAndEnforce` has two overloads as to what this argument can be, it can be either:
- A no-arg callable (that's what `DenyCallable` defines and why I casted here)
- A callable that takes a single argument of whatever type the resource passed to `queryAndEnforce` was (e.g. if we queried it with a `CatalogSchemaName` it'd pass that to the deny)

`denyShowSchemas` can be called with no-args or with a single string argument (`extraInfo`). In this instance we want to ensure the no-arg method is called, hence the cast.

I can alternatively change this for

```java
() -> AccessDeniedException.denyShowSchemas()
```

But it felt unnecessarily verbose.
What do you think?